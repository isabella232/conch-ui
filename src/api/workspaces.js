import axios from 'axios';

export const getAllRacks = (id) => {
    return axios({
        method: 'GET',
        url: `/workspace/${id}/rack`
    });
};

export const getDevices = (id) => {
    return axios({
        method: 'GET',
        url: `/workspace/${id}/device`
    });
};

export const getRackById = (id, rackId) => {
    return axios({
        method: 'GET',
        url: `/workspace/${id}/rack/${rackId}`
    })
    .then(response => {
        response.slots = response.slots.reduce((obj, curr) => {
            obj[curr.rack_unit_start] = curr;
            return obj;
        }, {});

        return Promise.resolve(response);
    });
};

export const loadAllWorkspaces = () => {
    return axios({
        method: 'GET',
        url: '/workspace'
    });
};

export const setRackLayout = (id, rackId, layout) => {
    return axios({
        method: 'POST',
        url: `/workspace/${id}/rack/${rackId}/layout`,
        data: layout
    });
};

export default {
    getAllRacks,
    getDevices,
    getRackById,
    loadAllWorkspaces,
    setRackLayout,
};
