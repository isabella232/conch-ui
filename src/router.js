import Vue from 'vue';
import Router from 'vue-router';
import Devices from './views/Devices/Devices.vue';
import DataCenterBrowser from './views/DataCenterBrowser/DataCenterBrowser.vue';
import SignIn from './views/SignIn/SignIn.vue';
import Status from './views/Status/Status.vue';
import UserProfile from './views/User/UserProfile.vue';
import Navbar from './views/Navbar/Navbar.vue';
import Sidebar from './views/Sidebar/Sidebar.vue';

Vue.use(Router);

export default new Router({
    routes: [
        {
            path: '/',
            name: 'signIn',
            component: SignIn,
        },
        {
            path: '/:currentWorkspace/status',
            name: 'status',
            components: {
                default: Status,
                sidebar: Sidebar,
                navbar: Navbar,
            },
            children: [
                {
                    // when /status/device/:deviceId is matched
                    path: 'device/:deviceId',
                    component: Status,
                },
            ],
        },
        {
            path: '/:currentWorkspace/device',
            name: 'workspace',
            components: {
                default: Devices,
                sidebar: Sidebar,
                navbar: Navbar,
            },
            children: [
                {
                    // when /device/:deviceId is matched
                    path: ':deviceId',
                    component: Devices,
                }
            ]
        },
        {
            path: '/:currentWorkspace/datacenter',
            name: 'datacenter',
            components: {
                default: DataCenterBrowser,
                sidebar: Sidebar,
                navbar: Navbar,
            },
            children: [
                {
                    // when /datacenter/:roomName/rack is matched
                    path: ':roomName/rack',
                    name: 'room',
                    component: DataCenterBrowser,
                    children: [
                        {
                            // when /datacenter/:roomName/rack/:rackId/device is matched
                            path: ':rackId/device',
                            name: 'rack',
                            component: DataCenterBrowser,
                            children: [
                                {
                                    // when /datacenter/:roomName/rack/:rackId/device/:deviceId is matched
                                    path: ':deviceId',
                                    name: 'device',
                                    component: DataCenterBrowser,
                                },
                            ],
                        },
                    ],
                },
            ],
        },
        {
            path: '/user',
            name: 'user',
            components: {
                default: UserProfile,
                sidebar: Sidebar,
                navbar: Navbar,
            },
        },
    ]
});
