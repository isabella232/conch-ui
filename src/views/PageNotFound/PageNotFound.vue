<template>
    <div class="page-not-found hero is-fullheight">
        <div class="content-left">
            <h1 class="title">404</h1>
            <h2 class="subtitle is-5">
                The page you're looking for is... missing.
            </h2>
            <hr />
            <p class="is-size-5">
                Head back to the last known workspace.
            </p>
            <p class="is-size-5">
                If you think something is broken, please report it.
            </p>
            <div class="buttons">
                <a
                    class="button is-outlined is-info is-uppercase has-text-weight-semibold"
                    @click="routeLastKnownWorkspace()"
                >
                    Last Known Workspace
                </a>
                <a
                    class="button is-outlined is-danger is-uppercase has-text-weight-semibold"
                    href="https://github.com/joyent/conch-ui/issues/new"
                    target="_blank"
                >
                    Report a Problem
                </a>
            </div>
        </div>
        <div class="content-right">
            <img src="../../assets/taken.svg" width="550" />
        </div>
    </div>
</template>

<script>
import { mapGetters, mapState } from 'vuex';

export default {
    methods: {
        routeLastKnownWorkspace() {
            let currentWorkspaceId = this.currentWorkspaceId;

            if (!currentWorkspaceId) {
                const routeParams = this.$route.params;

                if (routeParams && routeParams.currentWorkspace) {
                    this.loadCurrentWorkspace(routeParams.currentWorkspace);
                }
            }

            if (!currentWorkspaceId) {
                currentWorkspaceId = this.loadCurrentWorkspace;
            }

            if (!currentWorkspaceId) {
                currentWorkspaceId = this.workspaces[0].id;
            }

            this.$router.push({
                name: 'status',
                params: { currentWorkspace: currentWorkspaceId },
            });
        },
    },
    computed: {
        ...mapGetters(['currentWorkspaceId', 'loadCurrentWorkspace']),
        ...mapState(['workspaces']),
    },
};
</script>
