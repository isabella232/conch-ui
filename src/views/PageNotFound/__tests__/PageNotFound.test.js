import PageNotFound from '../PageNotFound.vue';
import VueRouter from 'vue-router';
import Vuex from 'vuex';
import router from '../../../router.js';
import { createLocalVue, shallowMount } from '@vue/test-utils';

// Fixture
import workspaces from '@src/__fixtures__/workspaces.js';

const localVue = createLocalVue();
localVue.use(VueRouter);
localVue.use(Vuex);

describe('PageNotFound.vue', () => {
    let getters;
    let state;
    let store;
    let wrapper;

    beforeEach(() => {
        getters = {
            currentWorkspaceId: jest.fn(),
            loadCurrentWorkspace: jest.fn(),
        };
        state = { workspaces };
        store = new Vuex.Store({ getters, state });
        wrapper = shallowMount(PageNotFound, { localVue, router, store });
    });

    // Helper function
    const clickWorkspaceButton = () => {
        wrapper
            .findAll('a.button')
            .at(0)
            .trigger('click');
    };

    test('should render correctly on initial render', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('should redirect user to status page when "Last Known Workspace" button clicked', () => {
        clickWorkspaceButton();

        expect(wrapper.vm.$route.name).toEqual('status');
        expect(wrapper.vm.$route.params).toHaveProperty('currentWorkspace');
    });

    test('should call routeLastKnownWorkspace when "Last Known Workspace" button clicked', () => {
        const spy = jest.spyOn(wrapper.vm, 'routeLastKnownWorkspace');

        // Needed in order for test to pass. It seems like Jest doesn't know
        // how to handle getters that take arguments. It seems to identify such
        // calls as method calls and looks for a corresponding item in the
        // 'methods' object of the Vue component.
        Object.defineProperty(wrapper.vm, 'loadCurrentWorkspace', {
            value: jest.fn(),
        });

        clickWorkspaceButton();

        expect(spy).toHaveBeenCalled();
    });
});
