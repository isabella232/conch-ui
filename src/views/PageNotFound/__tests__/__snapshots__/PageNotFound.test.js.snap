// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageNotFound.vue should render correctly on initial render 1`] = `
<div class="page-not-found">
  <div class="container">
    <div class="columns">
      <div class="column is-6 content">
        <h1 class="title">404</h1>
        <h2 class="subtitle is-5">
          The page you're looking for must be in another castle.
        </h2>
        <hr>
        <div class="action">
          <div class="text">
            <p class="is-size-5">
              Head back to the last known workspace.
            </p>
            <p class="is-size-5">
              If you think something is broken, please report it.
            </p>
          </div>
          <p class="buttons"><a class="button is-rounded is-outlined is-info is-uppercase has-text-weight-semibold">
              Last Known Workspace
            </a> <a href="https://github.com/joyent/conch-ui/issues/new" target="_blank" class="button is-rounded is-outlined is-danger is-uppercase has-text-weight-semibold">
              Report a Problem
            </a></p>
        </div>
      </div>
      <div class="column is-5 is-offset-1 image"><img src="../../assets/castle.svg" width="500"></div>
    </div>
  </div>
</div>
`;
